Source/WebCore/animation/CSSPropertyAnimation.cpp

```diff
@@ -4132,32 +4134,93 @@ bool CSSPropertyAnimation::animationOfPropertyIsAccelerated(AnimatableProperty p
     );
 }
 
-bool CSSPropertyAnimation::propertiesEqual(AnimatableProperty property, const RenderStyle& a, const RenderStyle& b)
+bool CSSPropertyAnimation::propertiesEqual(AnimatableProperty property, const RenderStyle& a, const RenderStyle& b, const Document& document)
 {
     return WTF::switchOn(property,
         [&] (CSSPropertyID propertyId) {
             if (auto* wrapper = CSSPropertyAnimationWrapperMap::singleton().wrapperForProperty(propertyId))
                 return wrapper->equals(a, b);
             return true;
-        }, [&] (const AtomString&) {
-            // FIXME: implement support for whether custom property values are equal.
-            ASSERT_NOT_REACHED();
+        }, [&] (const AtomString& customProperty) {
+            auto [aCustomPropertyValue, bCustomPropertyValue] = customPropertyValuesForBlending(&document, customProperty, a.getCustomProperty(customProperty), b.getCustomProperty(customProperty));
+            if (!aCustomPropertyValue && !bCustomPropertyValue)
+                return true;
+            if (!aCustomPropertyValue || !bCustomPropertyValue)
+                return false;
+
+            if (std::holds_alternative<CSSCustomPropertyValue::SyntaxValueList>(aCustomPropertyValue->value()) && std::holds_alternative<CSSCustomPropertyValue::SyntaxValueList>(bCustomPropertyValue->value()))
+                return std::get<CSSCustomPropertyValue::SyntaxValueList>(aCustomPropertyValue->value()) == std::get<CSSCustomPropertyValue::SyntaxValueList>(bCustomPropertyValue->value());
+
+            if (std::holds_alternative<CSSCustomPropertyValue::SyntaxValue>(aCustomPropertyValue->value()) && std::holds_alternative<CSSCustomPropertyValue::SyntaxValue>(bCustomPropertyValue->value()))
+                return std::get<CSSCustomPropertyValue::SyntaxValue>(aCustomPropertyValue->value()) == std::get<CSSCustomPropertyValue::SyntaxValue>(bCustomPropertyValue->value());
+
             return false;
         }
     );
 }
 
-bool CSSPropertyAnimation::canPropertyBeInterpolated(AnimatableProperty property, const RenderStyle& a, const RenderStyle& b)
+static bool typeOfSyntaxValueCanBeInterpolated(const CSSCustomPropertyValue::SyntaxValue& syntaxValue)
+{
+    return WTF::switchOn(syntaxValue,
+        [] (const Length&) {
+            return true;
+        },
+        [] (const StyleColor&) {
+            return true;
+        },
+        [] (CSSCustomPropertyValue::NumericSyntaxValue) {
+            return true;
+        },
+        [] (const CSSCustomPropertyValue::TransformSyntaxValue&) {
+            return true;
```

Comment:
could write these as single lines for compactness.

Reply:
@darinadler mentioned in a previous patch review that we preferred the vertical style. But I may have misunderstood what he meant by that.